## Проект с SQL

### Задания:

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Импорт библиотек

In [1]:
import pandas as pd
from sqlalchemy import create_engine 

### Подключение к БД

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db']) 

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

### Общие сведения о таблицах

In [4]:
sql= '''SELECT * FROM books'''
books = select(sql)

In [5]:
books.info()
books.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
sql= '''SELECT * FROM authors'''
authors = select(sql)

In [7]:
authors.info()
authors.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
sql= '''SELECT * FROM publishers'''
publishers = select(sql)

In [9]:
publishers.info()
publishers.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
sql= '''SELECT * FROM ratings'''
ratings = select(sql)

In [11]:
ratings.info()
ratings.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
sql= '''SELECT * FROM reviews'''
reviews = select(sql)

In [13]:
reviews.info()
reviews.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Создание запросов

#### 1. Cколько книг вышло после 1 января 2000 года

In [14]:
sql = '''SELECT count(distinct book_id) as count_books FROM books
        WHERE publication_date > '2000-01-01' '''

In [15]:
select(sql)

count_books
0          819

Из 1000 книг 819 вышли после 1 января 2000 года.

#### 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [16]:
sql = '''SELECT t.book_id,  count(distinct r.review_id) as count_review, avg(t.rating) as mean_rating
        FROM ratings t
        INNER JOIN reviews r on t.book_id = r.book_id
        GROUP BY t.book_id 
        ORDER by mean_rating '''

In [17]:
select(sql)

book_id  count_review  mean_rating
0        303             2         1.50
1        316             2         2.00
2        371             2         2.00
3        202             3         2.00
4        915             3         2.25
..       ...           ...          ...
989      459             2         5.00
990      699             2         5.00
991      213             2         5.00
992      441             2         5.00
993      390             2         5.00

[994 rows x 3 columns]

Всего обзоры есть на 994 книги из 1000, средняя оценка от 1,5 до 5.

#### 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [18]:
sql = '''SELECT count(t.book) as count_books, t.publisher FROM
        (SELECT b.book_id as book, b.num_pages, p.publisher as publisher
        FROM books b
        LEFT JOIN publishers p on b.publisher_id = p.publisher_id
        WHERE b.num_pages > 50
        GROUP BY b.book_id, p.publisher
        ) t
        GROUP BY t.publisher
        ORDER BY count_books desc
        LIMIT 1'''

In [19]:
select(sql)

count_books      publisher
0           42  Penguin Books

42 книги толще 50 страниц выпустил издатель Penguin Books.

#### 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [20]:
sql = ''' SELECT * FROM

        (SELECT a.author as author, avg(r.rating) as mean_rating from authors a
        LEFT JOIN books b on a.author_id = b.author_id
        LEFT JOIN ratings r on b.book_id = r.book_id
        
        WHERE b.book_id in (SELECT t.book_id FROM (SELECT book_id, count(rating_id) as count_rating FROM ratings GROUP BY book_id) t
                            WHERE t.count_rating > 50 GROUP BY book_id)
                        
        GROUP BY b.book_id, a.author) t
        
        
        WHERE t.mean_rating = 
        
        (SELECT max(t.mean_rating) FROM (SELECT b.book_id as book, a.author as author, avg(r.rating) as mean_rating from authors a
        LEFT JOIN books b on a.author_id = b.author_id
        LEFT JOIN ratings r on b.book_id = r.book_id
        
        WHERE b.book_id in (SELECT t.book_id FROM (SELECT book_id, count(rating_id) as count_rating FROM ratings GROUP BY book_id) t
                            WHERE t.count_rating > 50 GROUP BY book_id)
                        
        GROUP BY author, book) t )
        
        
        '''

In [21]:
select(sql)

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.414634

Вот тут я не согласна с записью, почему рядом с Роулинг Мэри ГрандПре? Это иллюстратор книг о Гарри Поттере, а не автор. Я бы иллюстратора в отдельном столбце хранила..

#### 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [22]:
sql = '''SELECT avg(count_review) as mean_count_review
                FROM
                (SELECT r.username as user, count(distinct r.review_id) as count_review, count(distinct s.rating_id) as count_rating FROM reviews r
                LEFT JOIN ratings s on r.username = s.username
                GROUP BY r.username) t
                
            WHERE t.count_rating > 50
            
            '''

In [23]:
select(sql)

mean_count_review
0          24.333333

В среднем 24 обзора у активных пользователей, которые ставили более 50 оценок.